grpc/transport-auth.go

```diff
@@ -0,0 +1,87 @@
+package grpc
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+)
+
+// transportCredentials is a grpc/credentials.TransportCredentials which supports
+// connecting to, and verifying multiple DNS names
+type transportCredentials struct {
+	configs map[string]*tls.Config
+}
+
+func newTransportCredentials(addrs []string, rootCAs *x509.CertPool, clientCerts []tls.Certificate) (credentials.TransportCredentials, error) {
+	configs := make(map[string]*tls.Config, len(addrs))
+	for _, addr := range addrs {
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		configs[addr] = &tls.Config{
+			ServerName:   host,
+			RootCAs:      rootCAs,
+			Certificates: clientCerts,
+		}
+	}
+	return &transportCredentials{configs}, nil
+}
+
+// ClientHandshake performs the TLS handshake for a client -> server connection
+func (tc *transportCredentials) ClientHandshake(addr string, rawConn net.Conn, timeout time.Duration) (net.Conn, credentials.AuthInfo, error) {
+	ctx := context.Background()
+	var cancel func()
+	if timeout != 0 {
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+	config, present := tc.configs[addr]
+	if !present {
+		return nil, nil, fmt.Errorf("boulder/grpc: Unexpected name, no TLS configuration present for \"%s\"", addr)
```

Comment:
Why not construct config objects inside `ClientHandshake` rather than looking them up? That way we could skip this whole error case. Note that it should be fine for `ClientHandshake` to be willing to connect to any hostname, because the only requested hostnames will be those explicitly configured (i.e., they should be trusted).